Inconsistency between adjoint gradients and finite differences

In Version 1.17.1, which is the latest version, adjoint gradients are inconsistent with finite differences when using `FourierFields` and `Near2FarFields`.

The tests are based on the tutorial code in this Github site. Take the tutorial code for [Fourier Fields Optimization in Waveguide Bend](https://github.com/NanoComp/meep/blob/master/python/examples/adjoint_optimization/Fourier%20Fields%20Optimization%20in%20Waveguide%20Bend.ipynb) as an example.
Some comparisons between adjoint gradients and finite differences are illustrated in the figure below, which displays obvious discrepancies. Increasing resolution does not help. If adjoint gradients were consistent with finite differences, the green dots would be close to the blue line (y=x), and the orange line (linear fitting of the green dots) would basically overlap with the blue line (y=x).
![image](https://user-images.githubusercontent.com/60903466/106056720-faa27100-60bc-11eb-9c95-cbfdb39f47b8.png)

The code for generating this figure is below:
```
x0 = 0.5*np.ones(Nx*Ny)
opt.update_design([x0])
f0, dJ_deps = opt()
db = 1e-3
choose = 5

g_discrete, idx = opt.calculate_fd_gradient(num_gradients=choose,db=db)
g_discrete = np.array(g_discrete)
g_adjoint = np.sum(dJ_deps, axis=1)

(m, b) = np.polyfit(np.squeeze(g_discrete), g_adjoint[idx], 1)
min_g = np.min(g_discrete)
max_g = np.max(g_discrete)

fig, ax1 = plt.subplots()
ax1.plot([min_g, max_g],[min_g, max_g],label='y=x comparison')
ax1.plot([min_g, max_g],[m*min_g+b, m*max_g+b],'--',label='Best fit m='+str(m))
ax1.plot(g_discrete,g_adjoint[idx],'o',label='Adjoint comparison')
ax1.set_xlabel('Finite Difference Gradient')
ax1.set_ylabel('Adjoint Gradient')
ax1.tick_params(axis='y')
plt.legend()
plt.grid(True)

g_discrete = g_discrete.reshape(-1, 1)
g_adjoint = g_adjoint.reshape(-1,1)
ax2 = ax1.twinx()
ax2.plot(g_discrete, np.abs((g_discrete-g_adjoint[idx])/g_discrete),'^',label='Relative error')
ax2.set_ylabel('Relative error')
ax2.tick_params(axis='y')

plt.title('Resolution: {}'.format(resolution))
plt.legend()
fig.tight_layout()
plt.show()
```
The code above is similar to that in the [introduction](https://github.com/NanoComp/meep/blob/master/python/examples/adjoint_optimization/01-Introduction.ipynb).

It seems that indeed #1456 is responsible for breaking the DFT fields objective function (`FourierFields`) used in the adjoint solver. I tested using a similar unit test for the adjoint solver that was added in #1472 which checks the gradient from the adjoint solver against the finite-difference result for a straight waveguide (script is below). As shown in the results below, the test passes prior to #1456 but fails with #1456 (the gradient changes by two orders of magnitude). I haven't tested the `Near2FarFields` objective function.

1. before #1456
```
directional_derivative:, -0.0000212148, -0.0000210261
.
----------------------------------------------------------------------
Ran 1 test in 49.983s

OK

```

2. after #1456
```
=====================================================================
FAIL: test_adjoint_solver (__main__.TestAdjointSolver)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "adjoint_solver_dft_fields.py", line 145, in test_adjoint_solver
    self.assertAlmostEqual(np.dot(dp,adjsol_grad),Ez2_perturbed-Ez2_unperturbed,places=6)
AssertionError: 2.4141666605149455e-07 != -2.1026055828266e-05 within 6 places
```

unit test
```py
import meep as mp
try:
    import meep.adjoint as mpa
except:
    import adjoint as mpa
import numpy as np
from autograd import numpy as npa
from autograd import tensor_jacobian_product
import unittest

np.random.seed(9861548)
resolution = 40

silicon = mp.Medium(epsilon=12)

sxy = 5.0
cell_size = mp.Vector3(sxy,sxy,0)

dpml = 1.0
boundary_layers = [mp.PML(thickness=dpml)]

eig_parity = mp.EVEN_Y+mp.ODD_Z

design_shape = mp.Vector3(1.5,1.5)
design_region_resolution = int(2*resolution)
Nx = int(design_region_resolution*design_shape.x)
Ny = int(design_region_resolution*design_shape.y)

w = 1.0 # waveguide width                                                                                                      
waveguide_geometry = [mp.Block(material=silicon,
                               center=mp.Vector3(),
                               size=mp.Vector3(mp.inf,w,mp.inf))]

fcen = 1/1.55
df = 0.2*fcen
sources = [mp.EigenModeSource(src=mp.GaussianSource(fcen,fwidth=df),
                              center=mp.Vector3(-0.5*sxy+dpml,0),
                              size=mp.Vector3(0,sxy,0),
                              eig_band=1,
                              eig_parity=eig_parity,
                              eig_match_freq=True)]

def forward_simulation(design_params):
    matgrid = mp.MaterialGrid(mp.Vector3(Nx,Ny),
                              mp.air,
                              silicon,
                              design_parameters=design_params.reshape(Nx,Ny),
                              grid_type='U_SUM')

    matgrid_geometry = [mp.Block(center=mp.Vector3(),
                                 size=mp.Vector3(design_shape.x,design_shape.y,0),
                                 material=matgrid)]

    geometry = waveguide_geometry + matgrid_geometry

    sim = mp.Simulation(resolution=resolution,
                        cell_size=cell_size,
                        boundary_layers=boundary_layers,
                        sources=sources,
                        geometry=geometry)

    mode = sim.add_dft_fields([mp.Ez],
                              fcen,
                              0,
                              1,
                              center=mp.Vector3(0.5*sxy-dpml),
                              size=mp.Vector3(0,sxy,0),
                              yee_grid=False)

    sim.run(until_after_sources=20)

    ez_dft = sim.get_dft_array(mode, mp.Ez, 0)
    ez2 = abs(ez_dft[61])**2

    sim.reset_meep()

    return ez2

def adjoint_solver(design_params):
    matgrid = mp.MaterialGrid(mp.Vector3(Nx,Ny),
                              mp.air,
                              silicon,
                              design_parameters=np.ones((Nx,Ny)))

    matgrid_region = mpa.DesignRegion(matgrid,
                                      volume=mp.Volume(center=mp.Vector3(),
                                                       size=mp.Vector3(design_shape.x,design_shape.y,0)))

    matgrid_geometry = [mp.Block(center=matgrid_region.center,
                                 size=matgrid_region.size,
                                 material=matgrid)]

    geometry = waveguide_geometry + matgrid_geometry

    sim = mp.Simulation(resolution=resolution,
                        cell_size=cell_size,
                        boundary_layers=boundary_layers,
                        sources=sources,
                        geometry=geometry)

    obj_list = [mpa.FourierFields(sim,
                                  mp.Volume(center=mp.Vector3(0.5*sxy-dpml),
                                            size=mp.Vector3(0,sxy,0)),
                                  mp.Ez)]

    def J(mode_mon):
        return npa.abs(mode_mon[0,61])**2

    opt = mpa.OptimizationProblem(
        simulation = sim,
        objective_functions = J,
        objective_arguments = obj_list,
        design_regions = [matgrid_region],
        frequencies=[fcen],
        decay_fields=[mp.Ez])

    f, dJ_du = opt([design_params])

    sim.reset_meep()

    return f, dJ_du

class TestAdjointSolver(unittest.TestCase):

    def test_adjoint_solver(self):
        p = np.random.rand(Nx*Ny)

        ## compute gradient using adjoint solver                                                                               
        adjsol_obj, adjsol_grad = adjoint_solver(p)

        ## compute unperturbed Ez2                                                                                             
        Ez2_unperturbed = forward_simulation(p)

        print("Ez2:, {:.6f}, {:.6f}".format(adjsol_obj,Ez2_unperturbed))
        self.assertAlmostEqual(adjsol_obj,Ez2_unperturbed,places=3)

        ## random epsilon perturbation for computing gradient via finite difference                                            
        deps = 1e-5
        dp = deps*np.random.rand(Nx*Ny)

        ## compute perturbed Ez2                                                                                               
        Ez2_perturbed = forward_simulation(p+dp)

        print("directional_derivative:, {:.10f}, {:.10f}".format(np.dot(dp,adjsol_grad),Ez2_perturbed-Ez2_unperturbed))
        self.assertAlmostEqual(np.dot(dp,adjsol_grad),Ez2_perturbed-Ez2_unperturbed,places=6)

if __name__ == '__main__':
    unittest.main()

```

--- a/src/array_slice.cpp
+++ b/src/array_slice.cpp
@@ -757,19 +757,13 @@ std::vector<double> fields::get_array_metadata(const volume &where) {
   vec min_max_loc[2]; // extremal points in subgrid
   int rank = get_array_slice_dimensions(where, dims, dirs, true, false, min_max_loc);
 
-  int full_rank = rank;
-  direction full_dirs[3];
-  for (int fr = 0; fr < rank; fr++)
-    full_dirs[fr] = dirs[fr];
-
   double *weights = get_array_slice(where, NO_COMPONENT);
 
   /* get length and endpoints of x,y,z tics arrays */
   size_t nxyz[3] = {1, 1, 1};
   double xyzmin[3] = {0.0, 0.0, 0.0}, xyzmax[3] = {0.0, 0.0, 0.0};
-  for (int fr = 0, rr = 0; fr < full_rank; fr++) {
-    direction d = full_dirs[fr];
-    int nd = d - X;
+  for (int nd = 0, rr = 0; nd < 3; ++nd) {
+    direction d = direction(nd);
     if (where.in_direction(d) == 0.0) {
       xyzmin[nd] = xyzmax[nd] = where.in_direction_min(d);
       nxyz[nd] = 1;

